File: client/src/modules/admin/models/address.ts
import type { AddressDraft, AddressSchema as BaseAddressSchema, AddressUpdates } from 'shared/types';
import { EMPTY_STRING, ZERO } from 'shared/constants';

export class AddressSchema implements BaseAddressSchema {
  public id: number;
  public country: string;
  public state: string;
  public city: string;
  public zipCode: string;
  public address: string;
  public createdAt: Date | null;
  public updatedAt: Date | null;

  private constructor(address?: BaseAddressSchema) {
    this.id = address?.id || ZERO;
    this.country = address?.country || EMPTY_STRING;
    this.state = address?.state || EMPTY_STRING;
    this.city = address?.city || EMPTY_STRING;
    this.zipCode = address?.zipCode || EMPTY_STRING;
    this.address = address?.address || EMPTY_STRING;
    this.createdAt = address?.createdAt ? new Date(address.createdAt) : null;
    this.updatedAt = address?.updatedAt ? new Date(address.updatedAt) : null;
  }

  public static create(address?: BaseAddressSchema): AddressSchema {
    return new AddressSchema(address);
  }

  public getUpdates(): AddressUpdates {
    const { id, createdAt, updatedAt, ...updates } = this;

    return updates;
  }
}

export function createAddressDraft(schema: BaseAddressSchema = AddressSchema.create()): AddressDraft {

  return {
    set country(country: string) {
      schema.country = country;
    },

    set state(state: string) {
      schema.state = state;
    },

    set city(city: string) {
      schema.city = city;
    },

    set zipCode(zipCode: string) {
      schema.zipCode = zipCode;
    },

    set address(address: string) {
      schema.address = address;
    },
  };

}

File: server/src/main/java/com/digitazon/poscoffee/models/Address.java
package com.digitazon.poscoffee.models;

import java.util.Date;

import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.Size;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import com.digitazon.poscoffee.shared.constants.AddressConstants;
import com.digitazon.poscoffee.shared.constants.AppConstants;

import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = AppConstants.DatabaseTable.ADDRESSES)
@EntityListeners(AuditingEntityListener.class)
@Data
@NoArgsConstructor
public class Address {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  @NotBlank(
    message = AddressConstants.COUNTRY_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToCreate.class
  )
  @Size(
    min = AppConstants.MIN_UPDATE_LENGTH,
    message = AddressConstants.COUNTRY_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToUpdate.class
  )
  private String country;

  @NotBlank(
    message = AddressConstants.STATE_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToCreate.class
  )
  @Size(
    min = AppConstants.MIN_UPDATE_LENGTH,
    message = AddressConstants.STATE_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToUpdate.class
  )
  private String state;

  @NotBlank(
    message = AddressConstants.CITY_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToCreate.class
  )
  @Size(
    min = AppConstants.MIN_UPDATE_LENGTH,
    message = AddressConstants.CITY_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToUpdate.class
  )
  private String city;

  @NotBlank(
    message = AddressConstants.ZIP_CODE_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToCreate.class
  )
  @Size(
    min = AppConstants.MIN_UPDATE_LENGTH,
    message = AddressConstants.ZIP_CODE_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToUpdate.class
  )
  private String zipCode;

  @NotBlank(
    message = AddressConstants.ADDRESS_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToCreate.class
  )
  @Size(
    min = AppConstants.MIN_UPDATE_LENGTH,
    message = AddressConstants.ADDRESS_EMPTY_MESSAGE,
    groups = AppConstants.ValidationGroups.ToUpdate.class
  )
  private String address;

  @CreatedDate
  private Date createdAt;

  @LastModifiedDate
  private Date updatedAt;

}
